File: supabase/functions/_shared/mulberry32.ts
/**
 * Canonical Mulberry32 32-bit PRNG implementation.
 * Guarantees bit-identical output to Flutter Dart client for any 32-bit seed.
 */
export class Mulberry32 {
  private state: number;

  constructor(seed: number) {
    this.state = (seed >>> 0);
  }

  /**
   * Returns unsigned 32-bit integer in range [0, 4294967295]
   */
  public nextUint32(): number {
    let t = (this.state += 0x6D2B79F5) | 0;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return (t ^ (t >>> 14)) >>> 0;
  }

  /**
   * Returns floating-point number strictly within [0.0, 1.0)
   */
  public nextFloat(): number {
    return this.nextUint32() / 4294967296.0;
  }

  /**
   * Returns integer within inclusive [min, max] range
   */
  public nextInt(min: number, max: number): number {
    return min + (this.nextUint32() % (max - min + 1));
  }
}

File: lib/core/utils/mulberry32.dart
class Mulberry32 {
  int _state;

  Mulberry32(int seed) : _state = seed & 0xFFFFFFFF;

  static int _imul(int a, int b) {
    final aLo = a & 0xFFFF;
    final aHi = (a >> 16) & 0xFFFF;
    final bLo = b & 0xFFFF;
    final bHi = (b >> 16) & 0xFFFF;
    return ((aLo * bLo) + (((aHi * bLo + aLo * bHi) & 0xFFFF) << 16)) &
        0xFFFFFFFF;
  }

  /// Returns unsigned 32-bit integer in [0, 4294967295]
  int nextUint32() {
    _state = (_state + 0x6D2B79F5) & 0xFFFFFFFF;
    var t = _state;
    t = _imul(t ^ (t >> 15), t | 1);
    t = (t ^ (t + _imul(t ^ (t >> 7), t | 61))) & 0xFFFFFFFF;
    return (t ^ (t >> 14)) & 0xFFFFFFFF;
  }

  /// Returns double in [0.0, 1.0)
  double nextFloat() {
    return nextUint32() / 4294967296.0;
  }

  /// Returns integer in range [min, max] inclusive
  int nextInt(int min, int max) {
    assert(min <= max, 'min ($min) must be <= max ($max)');
    return min + (nextUint32() % (max - min + 1));
  }
}
